brian2/codegen/generators/base.py

```diff
@@ -215,10 +216,27 @@ def has_repeated_indices(self, statements):
                          if index not in ('_idx', '0'))
         return not all_unique
 
+    _translate_cache = {}
     def translate(self, code, dtype):
         '''
         Translates an abstract code block into the target language.
         '''
+        cache_key = _hashable([code, dtype, self.variables,
+                               self.variable_indices,
+                               self.override_conditional_write,
+                               self.allows_scalar_write,
+                               self.iterate_all,
+                               self.template_name,
+                               self.name,
+                               self.owner.name,
+                               {var.owner.name
+                                for var in self.variables.itervalues()
+                                if getattr(var, 'owner', None) is not None},
+                               prefs,
+                               self.__class__])
```

Comment:
Another alternative. If there is some (potentially expensive) way of writing a program that keeps track of which variables a function call actually accesses, then we could use that as part of the test script to assert that it really does only access variables that are part of the key. Not sure how feasible this is.